timemachine/fe/mcgregor.py

```diff
@@ -292,31 +316,28 @@ def atom_map_pop(map_1_to_2, map_2_to_1, idx, jdx):
     map_2_to_1[jdx] = UNMAPPED
 
 
-def _graph_is_disconnected(g, mapped_nodes, demapped_nodes, unvisited_nodes):
-    mapped_nodes = set(mapped_nodes)
-    demapped_nodes = set(demapped_nodes)
-    unvisited_nodes = set(unvisited_nodes)
+def _graph_is_disconnected(g, node_states):
+    node_states = tuple(node_states)
 
-    # check intersections
-    assert mapped_nodes.union(demapped_nodes).union(unvisited_nodes) == set(range(g.n_vertices))
-    assert len(mapped_nodes.intersection(demapped_nodes)) == 0
-    assert len(mapped_nodes.intersection(unvisited_nodes)) == 0
-    assert len(demapped_nodes.intersection(unvisited_nodes)) == 0
+    source = -1
+    n_mapped_nodes = 0
+    for a_idx, state in enumerate(node_states):
+        if state == NODE_STATE_MAPPED:
+            source = a_idx
+            n_mapped_nodes += 1
 
-    all_possible_nodes = mapped_nodes.union(unvisited_nodes)
+    if source != -1:
+        return not g._verify_partially_connected(source, node_states, n_mapped_nodes)
 
-    if len(mapped_nodes) > 0:
-        sg_all_possible = g.nxg.subgraph(all_possible_nodes)
-        sg_ccs = nx.connected_components(sg_all_possible)
-        # see if all the mapped nodes belong to the same connected component
-        for cc in sg_ccs:
-            if set(mapped_nodes).issubset(cc):
-                return False
-        return True
     else:
         return False
 
 
+NODE_STATE_DEMAPPED = 0
```

Comment:
nit: Put the constants at the top of the files

Reply:
willfix

Reply:
I've renamed the constants to:

```
NODE_STATE_VISITED_DEMAPPED = 0 # VISITED AND DEMAPPED
NODE_STATE_VISITED_MAPPED = 1 # VISITED AND MAPPED
NODE_STATE_UNVISITED = 2 # UNVISITED
```